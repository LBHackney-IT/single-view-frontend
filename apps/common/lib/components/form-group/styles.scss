@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/govuk-frontend/govuk/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-label/label";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-hint/hint";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-error-message/error-message";

.lbh-form-group {
  label,
  legend {
    sup {
      color: lbh-colour("lbh-error");
    }
  }
}

fieldset.lbh-form-group legend {
  display: contents;
  margin-bottom: 0.375rem;
}
