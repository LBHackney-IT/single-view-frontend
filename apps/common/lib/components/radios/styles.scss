@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-radios/radios";
