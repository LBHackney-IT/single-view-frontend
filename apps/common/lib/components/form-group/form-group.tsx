import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
  useMemo,
} from "react";

import classNames from "classnames";

import { widthOverrides } from "../../utils";
import { TextArea } from "../text-area";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface FormGroupProps {
  id: string;
  label?: string;
  name?: string;
  hint?: string;
  error?: string | false;
  required?: boolean;
  children: ReactElement;
  override?: number;
}

export type FormGroupComponent = Polymorphic.ForwardRefComponent<"div", FormGroupProps>;

export const FormGroup: FormGroupComponent = forwardRef(function FormGroup(
  {
    as: FormGroupComp = "div",
    id,
    name,
    label,
    hint,
    error,
    required,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const formGroupClasses = classNames(
    "govuk-form-group",
    {
      "govuk-form-group--error": !!error,
    },
    "lbh-form-group",
    widthOverrides(override),
    className,
  );

  const describedBy = useMemo(() => {
    const classes: string[] = [];
    if (hint) {
      classes.push(`${id}-hint`);
    }
    if (error) {
      classes.push(`${id}-error`);
    }
    return classes.join(" ");
  }, [id, hint, error]);

  const LabelComp =
    typeof FormGroupComp === "string" && (FormGroupComp as string) === "fieldset"
      ? "legend"
      : "label";

  const formGroup = (
    <FormGroupComp ref={ref} id={id} className={formGroupClasses} {...props}>
      <LabelComp className="govuk-label lbh-label" htmlFor={`${id}-field`}>
        {label}
        {required ? <sup aria-hidden="true">*</sup> : ""}
      </LabelComp>
      {!!hint && (
        <span id={`${id}-hint`} className="govuk-hint lbh-hint">
          {hint}
        </span>
      )}
      {!!error && (
        <span id={`${id}-error`} className="govuk-error-message lbh-error-message">
          <span className="govuk-visually-hidden">Error:</span> {error}
        </span>
      )}
      {!!children &&
        Children.only(
          cloneElement(children, {
            id: `${id}-field`,
            name,
            required,
            error: !!error,
            "aria-describedby": describedBy || undefined,
            ...children.props,
          }),
        )}
    </FormGroupComp>
  );

  return isValidElement(children) && children.type === TextArea ? (
    <div className="govuk-character-count">{formGroup}</div>
  ) : (
    formGroup
  );
});
