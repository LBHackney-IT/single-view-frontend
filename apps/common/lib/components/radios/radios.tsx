import React, {
  Children,
  ComponentPropsWithoutRef,
  ReactNode,
  cloneElement,
  forwardRef,
  isValidElement,
  useEffect,
  useMemo,
  useRef,
} from "react";
import mergeRefs from "react-merge-refs";

import cn from "classnames";
import { Radios as RadiosJs } from "lbh-frontend";

import "./styles.scss";

export interface RadioProps extends ComponentPropsWithoutRef<"input"> {
  id: string;
  hint?: string;
  children: ReactNode;
  conditionalId?: string;
  error?: string;
}

export const Radio = forwardRef<HTMLInputElement, RadioProps>(function Radio(
  { id, className, type = "radio", hint, children, conditionalId, error, ...props },
  ref,
) {
  return (
    <div className={cn("govuk-radios__item", className)}>
      <input
        ref={ref}
        id={id}
        className="govuk-radios__input"
        type={type}
        aria-describedby={hint ? `${id}-hint` : undefined}
        data-aria-controls={conditionalId}
        {...props}
      />
      <label className="govuk-label govuk-radios__label" htmlFor={id}>
        {children}
      </label>
      {hint ? (
        <span id={`${id}-hint`} className="govuk-hint govuk-radios__hint">
          {hint}
        </span>
      ) : null}
    </div>
  );
});

export const RadioDivider = forwardRef<HTMLDivElement, ComponentPropsWithoutRef<"div">>(
  function RadioDivider(props, ref) {
    return <div ref={ref} className="govuk-radios__divider" {...props} />;
  },
);

export const RadioConditional = forwardRef<
  HTMLDivElement,
  ComponentPropsWithoutRef<"div">
>(function RadioConditional(props, ref) {
  return (
    <div
      ref={ref}
      className="govuk-radios__conditional govuk-radios__conditional--hidden"
      {...props}
    />
  );
});

export interface RadioGroupProps extends ComponentPropsWithoutRef<"div"> {
  variant?: "base" | "small";
  inline?: boolean;
  name?: string;
  error?: string;
  required?: boolean;
}

export const RadioGroup = forwardRef<HTMLDivElement, RadioGroupProps>(function RadioGroup(
  { variant = "base", inline = false, name, children, error, required, ...props },
  ref,
) {
  const localRef = useRef<HTMLDivElement>();

  useEffect(() => {
    /* istanbul ignore else */
    if (localRef.current) {
      new RadiosJs(localRef.current).init();
    }
  }, []);

  const hasConditionals = useMemo(
    () =>
      Children.toArray(children).some(
        (child) => isValidElement(child) && child.type === RadioConditional,
      ),
    [children],
  );

  return (
    <div
      ref={mergeRefs([localRef, ref])}
      className={cn(
        "govuk-radios",
        {
          "govuk-radios--small": variant === "small",
          "govuk-radios--inline": inline,
          "govuk-radios--conditionals": hasConditionals,
        },
        "lbh-radios",
      )}
      {...props}
    >
      {Children.map(
        children,
        (child) =>
          child &&
          isValidElement(child) &&
          cloneElement(child, {
            name,
            required,
            ...child.props,
          }),
      )}
    </div>
  );
});
