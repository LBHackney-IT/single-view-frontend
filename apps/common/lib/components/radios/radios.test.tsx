import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";

import { FormGroup } from "../form-group";
import { Radio, RadioConditional, RadioDivider, RadioGroup } from "./radios";

test("it passes a11y on single radio", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="Radio Test">
      <Radio id="radio">Label</Radio>
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it passes a11y on radio group", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="Radios Test">
      <RadioGroup>
        <Radio id="radio-1">Label</Radio>
        <Radio id="radio-2">Label2</Radio>
      </RadioGroup>
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders a radio", () => {
  const { container } = render(<Radio id="test">Label</Radio>);
  expect(container).toMatchSnapshot();
});

test("it renders a radio with a hint", () => {
  const { container } = render(
    <Radio id="test" hint="A short description">
      Label
    </Radio>,
  );
  expect(container).toMatchSnapshot();
});

test("it renders a group of radios", () => {
  const { container } = render(
    <RadioGroup>
      <Radio id="test">Label</Radio>
      <Radio id="test2">Label2</Radio>
    </RadioGroup>,
  );
  expect(container).toMatchSnapshot();
});

test("it renders a group of radios with a divider", () => {
  const { container } = render(
    <RadioGroup>
      <Radio id="test">Label</Radio>
      <RadioDivider>Or</RadioDivider>
      <Radio id="test2">Label2</Radio>
    </RadioGroup>,
  );
  expect(container).toMatchSnapshot();
});

test("it renders a group of radios with a conditional", () => {
  const { container } = render(
    <RadioGroup>
      <Radio id="test" conditionalId="conditional">
        Label
      </Radio>
      <RadioConditional id="conditional">Hello</RadioConditional>
      <Radio id="test2">Label2</Radio>
    </RadioGroup>,
  );
  expect(container).toMatchSnapshot();
});
