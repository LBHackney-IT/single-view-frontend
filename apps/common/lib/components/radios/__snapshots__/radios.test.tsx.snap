// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y on radio group 1`] = `
<div>
  <div
    class="govuk-form-group lbh-form-group"
    id="test"
  >
    <label
      class="govuk-label lbh-label"
      for="test-field"
    >
      Radios Test
      
    </label>
    <div
      class="govuk-radios lbh-radios"
      id="test-field"
    >
      <div
        class="govuk-radios__item"
      >
        <input
          class="govuk-radios__input"
          id="radio-1"
          name="test"
          type="radio"
        />
        <label
          class="govuk-label govuk-radios__label"
          for="radio-1"
        >
          Label
        </label>
      </div>
      <div
        class="govuk-radios__item"
      >
        <input
          class="govuk-radios__input"
          id="radio-2"
          name="test"
          type="radio"
        />
        <label
          class="govuk-label govuk-radios__label"
          for="radio-2"
        >
          Label2
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`it passes a11y on single radio 1`] = `
<div>
  <div
    class="govuk-form-group lbh-form-group"
    id="test"
  >
    <label
      class="govuk-label lbh-label"
      for="test-field"
    >
      Radio Test
      
    </label>
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="radio"
        name="test"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="radio"
      >
        Label
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a group of radios 1`] = `
<div>
  <div
    class="govuk-radios lbh-radios"
  >
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test2"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a group of radios with a conditional 1`] = `
<div>
  <div
    class="govuk-radios govuk-radios--conditionals lbh-radios"
  >
    <div
      class="govuk-radios__item"
    >
      <input
        aria-controls="conditional"
        aria-expanded="false"
        class="govuk-radios__input"
        id="test"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-radios__conditional govuk-radios__conditional--hidden"
      id="conditional"
    >
      Hello
    </div>
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test2"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a group of radios with a divider 1`] = `
<div>
  <div
    class="govuk-radios lbh-radios"
  >
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-radios__divider"
    >
      Or
    </div>
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test2"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a radio 1`] = `
<div>
  <div
    class="govuk-radios__item"
  >
    <input
      class="govuk-radios__input"
      id="test"
      type="radio"
    />
    <label
      class="govuk-label govuk-radios__label"
      for="test"
    >
      Label
    </label>
  </div>
</div>
`;

exports[`it renders a radio with a hint 1`] = `
<div>
  <div
    class="govuk-radios__item"
  >
    <input
      aria-describedby="test-hint"
      class="govuk-radios__input"
      id="test"
      type="radio"
    />
    <label
      class="govuk-label govuk-radios__label"
      for="test"
    >
      Label
    </label>
    <span
      class="govuk-hint govuk-radios__hint"
      id="test-hint"
    >
      A short description
    </span>
  </div>
</div>
`;
